vendor/github.com/OpenBazaar/go-ethwallet/wallet/wallet.go

```diff
@@ -809,16 +706,22 @@ func (wallet *EthereumWallet) CheckTxnRcpt(hash *common.Hash, data []byte) (*com
 			if err != nil {
 				return nil, err
 			}
-			wallet.db.Txns().Delete(chash)
-			toAddr := common.HexToAddress(util.EnsureCorrectPrefix(pTxn.To))
+			err = wallet.db.Txns().Delete(chash)
 			n := new(big.Int)
 			n, _ = n.SetString(pTxn.Amount, 10)
-			go wallet.AssociateTransactionWithOrder(
-				wallet.createTxnCallback(util.EnsureCorrectPrefix(hash.Hex()), pTxn.OrderID, EthAddress{&toAddr},
-					*n, time.Now(), pTxn.WithInput))
+			if pTxn.Amount != "0" {
+				toAddr := common.HexToAddress(util.EnsureCorrectPrefix(pTxn.To))
+				go wallet.AssociateTransactionWithOrder(
+					wallet.createTxnCallback(util.EnsureCorrectPrefix(hash.Hex()), pTxn.OrderID, EthAddress{&toAddr},
+						*n, time.Now(), pTxn.WithInput))
+			} else {
+				addrScrHash := common.HexToAddress(pTxn.To)
+				go wallet.AssociateTransactionWithOrder(wallet.createTxnCallback(util.EnsureCorrectPrefix(hash.Hex()),
+					pTxn.OrderID, EthAddress{address: &addrScrHash}, *n,
+					time.Now(), true))
+			}
```

Comment:
Don't repeat the `AssociateTransactionWithOrder` in both branches. Instead, make the change that `if pTxn.Amount != "0" {` is interested in cleared by adjusting a `var hasInput bool` value can calling the whole method after the branch.

Reply:
I'm generally concerned that `AssociateTransactionWithOrder` can run indefinitely here and there'd be no way for us to clean this goroutine up if we wanted to stop the parent threads.